R/prevalence-DELTA_test.R

```diff
@@ -1,701 +1,1136 @@
+    row <- dplyr::bind_cols(
+      st[, c("rank", "feature", "group_col", "group1", "group2", "design")],
+      tibble::tibble(
+        n_subjects_paired = if (identical(st$design, "paired")) {
+          as.integer(P)
+        } else {
+          NA_integer_
+        },
+        n_peptides_used = as.integer(res$n_peptides_used),
+        m_eff = as.numeric(res$m_eff),
+        T_obs = as.numeric(res$T_obs),
+        T_obs_stand = {
+          sd_null <- as.numeric(res$T_null_sd)
+          ifelse(!is.na(sd_null) & sd_null > 0,
+            as.numeric(res$T_obs) / sd_null,
+            NA_real_
+          )
+        },
+        Z_from_p = {
+          pval <- as.numeric(res$p_perm)
+          T_val <- as.numeric(res$T_obs)
+          ifelse(
+            is.na(pval),
+            NA_real_,
+            sign(T_val) * stats::qnorm(pval / 2, lower.tail = FALSE)
+          )
+        },
+        b = as.integer(res$b),
+        p_perm = as.numeric(res$p_perm),
+        mean_delta = as.numeric(res$mean_delta),
+        frac_delta_pos = as.numeric(res$frac_delta_pos),
+        mean_delta_w = as.numeric(res$mean_delta_w),
+        frac_delta_pos_w = as.numeric(res$frac_delta_pos_w),
+        !!paste0("fold_change_", fold_change) := fc_val,
+        !!paste0("cross_prev_", cross_prev) := cp_val
+      )
     )
-    if (is.null(stream_path)) {
-      result_rows <- purrr::flatten(result_lists)
-    } else {
-      result_rows <- NULL # already streamed by workers
-    }
 
-    # master: log final percentage
-    if (isTRUE(log) && work_total > 0) {
-      conr <- file(.progress_file(log_file, stream_path), open = "rb")
-      done_w <- tryCatch(unserialize(conr), error = function(e) 0)
-      close(conr)
-      prog <- 100 * (done_w / work_total)
+    if (isTRUE(log)) {
+      msg <- sprintf(
+        "%s / %s | n_pep=%d b=%d p=%g",
+        row$rank, row$feature, row$n_peptides_used, row$b, row$p_perm
+      )
+      if (log_to_file) .ph_log_info_file(log_file, msg) else .ph_log_info(msg)
+
+      inc <- as.numeric(row$n_peptides_used) * B_num
+      done_w <- .progress_add(progress_path, inc)
+
+      prog <- if (is.finite(done_w) && is.finite(work_total) &&
+        work_total > 0) {
+        pmin(100, 100 * (done_w / work_total))
+      } else {
+        NA_real_
+      }
+
       if (log_to_file) {
-        .ph_log_ok_file(log_file, sprintf("progress=%.1f%% (final)", prog))
+        .ph_log_info_file(log_file, sprintf(
+          "progress=%s [C]",
+          if (is.na(prog)) "n/a" else sprintf("%.1f%%", prog)
+        ))
       } else {
-        .ph_log_ok(sprintf("progress=%.1f%% (final)", prog))
+        .ph_log_info(sprintf(
+          "progress=%s [C]",
+          if (is.na(prog)) "n/a" else sprintf("%.1f%%", prog)
+        ))
       }
     }
+
+    row
+  }
+
+  # --- 7/8) Computation path: sequential vs parallel --------------------------
```

Comment:
Note: I didn't read most of the function as it's very long and I don't fully understand the delta framework yet.
I trust you that the implementation works 😬